### PR TITLE
CORCI-938 build: Adding MACSio RPM repository

### PR DESCRIPTION
**Pull request by phender**
Adding repository for the RPM packaging of MACSio as a requirement for
adding MACSio to DAOS functional testing.

Signed-off-by: Phillip Henderson <phillip.henderson@intel.com>

### CODE REVIEW

File: MACSio.spec
Comment (brianjmurrell) [inline MACSio.spec]: This should be required by the `hdf5-mpich-devel` below, so not really needed here.
```suggestion
```

File: MACSio.spec
Comment (brianjmurrell) [inline MACSio.spec]: I don't think you need the `/3.3` here and having it just means have to keep playing whack-a-mole as newer versions of gnu-mpich are available.
```suggestion
  module load gnu-mpich
```

File: MACSio.spec
Comment (brianjmurrell) [inline MACSio.spec:58]: Surprising `/usr` isn't the default auto-probed default for this.

File: Makefile
Comment (brianjmurrell) [inline Makefile]: I guess this seems like it could/should be the minimum/default test we could use in `packaging/` for RPM builds when a repo doesn't define something more extensive.  Feel free to add that to [daos-stack/packaging](https://github.com/daos-stack/packaging) if you like.  :wink:

File: MACSio.spec
Comment (brianjmurrell) [inline MACSio.spec]: These are causing linting errors.  Even though you are never supposed to, it seems inconsistent when you can have macros in comments and when you can't in terms of when it's a linting warning vs. error.  These ones are producing an error.  When I run into this but want to keep the comment I usually just `s/%/#/g` one the line.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: MACSio.spec
Comment (phender) [reply MACSio.spec]: Removed.

File: MACSio.spec
Comment (phender) [reply MACSio.spec]: These lines are not required and have been removed.

**Comment by brianjmurrell** [comment]
FWIW, if you want to use daos-stack/hdf5#6 with this PR, as (a) an effort to test that it fixes the problem and (b) not have to wait for it to land, you can add a:

```
PR-repos: hdf5@PR-6
```

to the commit message of the last commit in this PR to use hdf5 in that PR instead of the one in the general repo.

**Comment by phender** [comment]
> FWIW, if you want to use [daos-stack/hdf5#6](https://github.com/daos-stack/hdf5/pull/6) with this PR, as (a) an effort to test that it fixes the problem and (b) not have to wait for it to land, you can add a:
> 
> ```
> PR-repos: hdf5@PR-6
> ```
> 
> to the commit message of the last commit in this PR to use hdf5 in that PR instead of the one in the general repo.

Exactly what I was I hoping we could do!  Is there something else required?  When I tried including the pragma the Leap 15 install still tried to install the older hdf5 rpm:
```
[2020-06-22T18:25:07.481Z]   - nothing provides gcc-gfortran(x86-64) needed by hdf5-devel-1.10.5-7.g07066a381e.suse.lp151.x86_64
```

File: Makefile
Comment (brianjmurrell) [inline Makefile]: As noted, this should come from an updated daos-stack/packaging, once daos-stack/packaging#30 lands.

File: MACSio.spec
Comment (brianjmurrell) [inline MACSio.spec]: Indeed.  You should always do a double-take when you see a BR: that is not a `-devel`.  It's certainly not a hard and fast rule, but if it's for a library, it should almost always be a `-devel`.

**Comment by brianjmurrell** [comment]
> Exactly what I was I hoping we could do! Is there something else required?

Heh.  Yeah, sadly: daos-stack/packaging#30.  You could go review that if you are comfortable with it.  That one has been sitting on the pile for a while.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: MACSio.spec
Comment (brianjmurrell) [inline MACSio.spec]: We're actually standardising on _lua-lmod_.  There is a PR for adding it to our repos in daos-stack/lua-lmod#1. I have not landed it yet as it's still being validated with the bigger body of work that is aimed at running Leap 15 in CI.  You can use `PR-repos: lua-lmod@PR-1` while it's still in a PR-only state.

This all could revert back to _Modules_ at some point in the future, if/when somebody has time to make it work up and down the entire CI stack, but I had to give up on that as it was becoming a time-suck and building a _lua-lmod_ was just the easier path.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: MACSio.spec
Comment (phender) [reply MACSio.spec]: I ran into this problem after applying [packaging(#30)](https://github.com/daos-stack/packaging/pull/30) and using `PR-repos: lua-lmod@PR-1`:
```
[2020-06-23T13:39:02.965Z] hdf5:PR-6:lastSuccessfulBuild                    82 kB/s |  13 kB     00:00    
[2020-06-23T13:39:02.965Z] lua-lmod:PR-1:lastSuccessfulBuild                11 kB/s | 546  B     00:00    
[2020-06-23T13:39:02.965Z] Failed to download metadata for repo 'lua-lmod:PR-1:lastSuccessfulBuild'
[2020-06-23T13:39:02.965Z] Error: Failed to download metadata for repo 'lua-lmod:PR-1:lastSuccessfulBuild'
```

Comment (phender) [reply MACSio.spec]: As a quick test I ran [build 16](https://build.hpdd.intel.com/job/daos-stack/job/MACSio/job/PR-1/16/) without `PR-repos: lua-lmod@PR-1` and re-enabled the `BuildRequires:  Modules` entry in the  spec file.  This confirmed the packaging changes work.

Regardless of Module vs. lua-lmod, I still have a problem with the Leap 15 build where it appears to have issue with the `hdf5-devel` provided `H5pubconf.h` which is renamed by the RPM to be `H5pubconf-64.h`.  This renaming is not a problem on the CentOS 7 build, but it appears to break the Leap 15 build:
```
[ 92%] Building CXX object macsio/CMakeFiles/macsio.dir/__/plugins/macsio_hdf5.c.o
/builddir/build/BUILD/MACSio-1.1/plugins/macsio_hdf5.c:50:10: fatal error: H5pubconf.h: No such file or directory
 #include <H5pubconf.h>
          ^~~~~~~~~~~~~
compilation terminated.
```
Is this a common issue with Leap 15 with a known workaround or is it something specific `hdf5-devel` is introducing that requires modify files or adding a link specific to Leap 15?

**Comment by phender** [comment]
Using "PR-repos-leap15: lua-lmod:PR-1" with the added support for "PR-repos-leap15" has resolved the issue with only being able to either OS.

File: packaging/Makefile_packaging.mk
Comment (brianjmurrell) [inline packaging/Makefile_packaging.mk]: You don't want to run these through this loop, but rather do the:

```bash
	rm -f /etc/yum.repos.d/*"${DAOS_STACK_LOCAL_REPO//\//_}"		
￼	yum-config-manager --add-repo="$REPOSITORY_URL"/"$DAOS_STACK_LOCAL_REPO"		
￼	echo "gpgcheck = False" >> /etc/yum.repos.d/*"${DAOS_STACK_LOCAL_REPO//\//_}".repo
```

process with them.  But just be aware that in the above, `DAOS_STACK_LOCAL_REPO` is just the path portion of the URL and `$($(DISTRO_BASE)_LOCAL_REPOS)` will be entire URLs, so you will have to extract the path portion out.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

**Comment by phender** [comment]
Verified the default Makefile test case changes work in https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2FMACSio/detail/PR-1/37/pipeline

**Review by brianjmurrell — APPROVED** [review]

**Review by brianjmurrell — APPROVED** [review]

**Comment by phender** [comment]
This is ready to be merged.

**Comment by brianjmurrell** [comment]
> This is ready to be merged.

Is this affected by daos-stack/hdf5#7 in any way?  Worth a run with `PR-repos: hdf5@PR-7` perhaps?

**Comment by phender** [comment]
> > This is ready to be merged.
> 
> Is this affected by [daos-stack/hdf5#7](https://github.com/daos-stack/hdf5/pull/7) in any way? Worth a run with `PR-repos: hdf5@PR-7` perhaps?

It should not be affected by hdf5#7 as that only enables building the hdf5-openmpi* RPMs, which this version does not use.  The next change, and why I'd like to merge this first, is to build a MACSio-openmpi package to make use of the hdf5-openmpi-devel RPM.

**Comment by brianjmurrell** [comment]
> It should not be affected by hdf5#7 as that only enables building the hdf5-openmpi* RPMs, which this version does not use. The next change, and why I'd like to merge this first, is to build a MACSio-openmpi package to make use of the hdf5-openmpi-devel RPM.

It also does some renaming of packages to have the mpich variant in their name.  I don't recall off-hand how many it does that with though.

Ultimately, any breakage will be found eventually, and given that this package is not used anywhere yet, landing it as is won't break anything.  It might just get broken on some future update.

**Comment by phender** [comment]
> It also does some renaming of packages to have the mpich variant in their name. I don't recall off-hand how many it does that with though.
>
Only the `hdf5-tests` RPM is renamed to `hdf5-mpich-tests`, which this RPM does not utilize.

> 
> Ultimately, any breakage will be found eventually, and given that this package is not used anywhere yet, landing it as is won't break anything. It might just get broken on some future update.
>
Right, and there is a planned update after this PR and PR-hdf5#7 are both merged, so any issues can be addressed then.

**Pull request merged**
